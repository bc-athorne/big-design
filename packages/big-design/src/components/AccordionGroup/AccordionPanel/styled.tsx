import { theme as defaultTheme, remCalc } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { Box } from '../../Box';
import { StyleableButton } from '../../Button/private';
import { GridItem } from '../../Grid/Item';

import { AccordionPanelProps } from './index';

export const StyledAccordionButton = styled(StyleableButton)`
  background: white;
  border-bottom: ${({ theme }) => theme.border.box};
  border-radius: 0;
  padding: ${({ theme }) => theme.spacing.xLarge};
  text-align: left;
  width: 100%;
  & > span {
    width: 100%;
  }
`;

export const StyledGridItem = styled(GridItem)`
  &:last-child {
    grid-column: 5;
    text-align: right;
  }
`;

export const StyledBox = styled(Box)<AccordionPanelProps>`
  border-bottom: ${({ theme }) => theme.border.box};
  padding: ${({ theme }) => theme.spacing.xLarge}};
  padding-left: ${({ iconLeft, theme }) => (iconLeft ? remCalc(60) : `${theme.spacing.xLarge}`)};
`;

export const StyledAccordion = styled(Box)<AccordionPanelProps>`
  &:last-child ${StyledAccordionButton} {
    border-bottom: ${({ isExpanded, theme }) =>
      isExpanded ? `${theme.border.box}` : `${theme.border.none}`};
    border-bottom-left-radius: ${({ isExpanded, theme }) =>
      isExpanded ? `${theme.borderRadius.none}` : `${theme.borderRadius.normal}`};
    border-bottom-right-radius: ${({ isExpanded, theme }) =>
      isExpanded ? `${theme.borderRadius.none}` : `${theme.borderRadius.normal}`};
  }
`;

StyledAccordionButton.defaultProps = { theme: defaultTheme };
StyledAccordion.defaultProps = { theme: defaultTheme };
StyledBox.defaultProps = { theme: defaultTheme };
StyledGridItem.defaultProps = { theme: defaultTheme };
