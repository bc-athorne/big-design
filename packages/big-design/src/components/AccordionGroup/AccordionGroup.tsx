import React, { memo } from 'react';

import { StyledAccordionGroup, StyledH2 } from './styled';

export interface AccordionGroupProps {
  header: string;
  children?: React.ReactNode;
}

export const AccordionGroup: React.FC<AccordionGroupProps> = memo(({ header, children }) => {
  return (
    <StyledAccordionGroup>
      <StyledH2>{header}</StyledH2>
      {children}
    </StyledAccordionGroup>
  );
});
