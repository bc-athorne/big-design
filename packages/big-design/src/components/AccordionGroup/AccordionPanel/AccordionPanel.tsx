import { Text } from '@bigcommerce/big-design';
import { ExpandLessIcon, ExpandMoreIcon } from '@bigcommerce/big-design-icons';
import React, { HTMLAttributes, memo } from 'react';

import { useUniqueId } from '../../../hooks';
import { BoxProps } from '../../Box/Box';
import { useAccordion } from '../hooks/useAccordion';

import { StyledAccordion, StyledAccordionButton, StyledBox, StyledGridItem } from './styled';

export interface AccordionPanelProps extends HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  header: string;
  isExpanded: boolean;
  onClick?: () => void;
  iconLeft?: React.ReactNode;
  defaultExpanded?: boolean;
}

export const AccordionPanel: React.FC<AccordionPanelProps & BoxProps> = memo(
  ({ children, header, iconLeft, defaultExpanded = false, ...props }) => {
    const { isExpanded, onClick } = useAccordion({ defaultExpanded });
    const accordionId = useUniqueId('accordion');
    const accordionItemId = useUniqueId('accordion-item');

    return (
      <StyledAccordion {...props} isExpanded={isExpanded}>
        <StyledAccordionButton
          aria-controls={accordionItemId}
          aria-expanded={isExpanded}
          id={accordionId}
          onClick={onClick}
          variant="subtle"
        >
          <StyledGridItem>
            {iconLeft}
            <Text as="span" marginLeft={iconLeft ? 'small' : 'none'}>
              {header}
            </Text>
          </StyledGridItem>
          <StyledGridItem>
            {isExpanded ? (
              <ExpandLessIcon color="secondary70" />
            ) : (
              <ExpandMoreIcon color="secondary70" />
            )}
          </StyledGridItem>
        </StyledAccordionButton>
        <StyledBox
          {...props}
          aria-labelledby={accordionId}
          display={isExpanded ? 'block' : 'none'}
          hidden={!isExpanded}
          iconLeft={iconLeft}
          id={accordionItemId}
          role="region"
        >
          {children}
        </StyledBox>
      </StyledAccordion>
    );
  },
);
