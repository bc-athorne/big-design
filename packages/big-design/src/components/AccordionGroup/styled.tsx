import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { Box } from '../Box';
import { StyleableH2 } from '../Typography/Typography';

export const StyledAccordionGroup = styled(Box)`
  border-radius: ${({ theme }) => theme.borderRadius.normal};
  border: ${({ theme }) => theme.border.box};
  ${({ theme }) => theme.shadow.raised};
`;

export const StyledH2 = styled(StyleableH2)`
  padding: ${({ theme }) => theme.spacing.xLarge};
  border-bottom: ${({ theme }) => theme.border.box};
  margin: 0;
`;

StyledAccordionGroup.defaultProps = { theme: defaultTheme };
