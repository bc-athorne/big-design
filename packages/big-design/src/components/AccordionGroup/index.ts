import { AccordionGroupProps as _AccordionGroupProps } from './AccordionGroup';

export { AccordionGroup } from './AccordionGroup';
export type AccordionGroupProps = _AccordionGroupProps;
