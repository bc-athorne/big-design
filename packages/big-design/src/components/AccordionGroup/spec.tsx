import { ErrorIcon } from '@bigcommerce/big-design-icons';
import userEvent from '@testing-library/user-event';
import React from 'react';

import { render, screen, waitFor } from '@test/utils';

import { Text } from '../Typography';

import { AccordionGroup } from './AccordionGroup';
import { AccordionPanel } from './AccordionPanel';

const handleClick = jest.fn();

const mockAccordion = (
  <AccordionPanel header="Panel Header" isExpanded onClick={handleClick}>
    <Text>This is a child component</Text>
  </AccordionPanel>
);

const mockAccordionGroup = (
  <AccordionGroup header="Test Accordion">
    <AccordionPanel header="Panel Header" isExpanded onClick={handleClick}>
      <Text>This is a child component</Text>
    </AccordionPanel>
  </AccordionGroup>
);

test('it renders accordion group header', async () => {
  render(mockAccordionGroup);

  const header = await screen.findByRole('heading', { name: /test accordion/i });

  expect(header).toMatchSnapshot();
  expect(header).toBeVisible();
});

test('it renders accordion children', async () => {
  render(mockAccordionGroup);

  const child = await screen.findByText('Panel Header');

  expect(child).toBeVisible();
});

test('accordion item renders header', async () => {
  render(mockAccordion);

  const header = await screen.findByText('Panel Header');

  expect(header).toBeVisible();
});

test('accordion row collapses and expands on click', async () => {
  render(mockAccordion);

  const panel = await screen.findByRole('button');

  userEvent.click(panel);

  await waitFor(() => {
    expect(panel.getAttribute('aria-expanded')).toBe('true');
  });

  userEvent.click(panel);

  await waitFor(() => {
    expect(panel.getAttribute('aria-expanded')).toBe('false');
  });
});

test('accordion row renders children', async () => {
  render(mockAccordion);

  const panel = await screen.findByRole('button');

  userEvent.click(panel);

  const child = await screen.findByRole('region');

  expect(child).toBeVisible();
});

test('if defaultExpanded is set to true, accordion is expanded', async () => {
  render(
    <AccordionPanel defaultExpanded={true} header="Panel Header" isExpanded onClick={handleClick}>
      <Text>This is a child component</Text>
    </AccordionPanel>,
  );

  const panel = await screen.findByRole('button');

  expect(panel.getAttribute('aria-expanded')).toBe('true');
});

test('it renders icon if iconLeft prop is set', async () => {
  render(
    <AccordionPanel
      defaultExpanded={true}
      header="Panel Header"
      iconLeft={<ErrorIcon />}
      isExpanded
      onClick={handleClick}
    >
      <Text>This is a child component</Text>
    </AccordionPanel>,
  );

  const icons = (await screen.findByRole('button')).querySelectorAll('svg');

  const icon = icons[0];

  expect(icon).toBeVisible();
});
