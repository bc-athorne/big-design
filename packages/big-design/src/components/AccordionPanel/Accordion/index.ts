import { AccordionPanelProps as _AccordionPanelProps } from './AccordionPanel';

export { AccordionPanel } from './AccordionPanel';
export type AccordionPanelProps = _AccordionPanelProps;
