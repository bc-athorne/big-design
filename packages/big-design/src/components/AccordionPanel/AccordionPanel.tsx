import React, { memo } from 'react';

import { Panel } from '../Panel';

import { AccordionPanel, AccordionPanelProps } from './AccordionPanel';
import { StyledAccordionPanelWrapper } from './styled';

export interface AccordionGroupProps {
  header: string;
  panels: AccordionPanelProps[];
}

export const AccordionGroup: React.FC<AccordionGroupProps> = memo(({ header, panels }) => {
  const renderedPanels = panels.map((panel, index) => <AccordionPanel {...panel} key={index} />);

  return (
    <Panel header={header}>
      <StyledAccordionPanelWrapper>{renderedPanels}</StyledAccordionPanelWrapper>
    </Panel>
  );
});
