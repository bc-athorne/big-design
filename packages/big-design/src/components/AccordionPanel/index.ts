import { AccordionPanelProps as _AccordionPanelProps } from './AccordionPanel';

export { AccordionPanel } from './AccordionPanel';
export { useAccordionPanel } from './hooks';
export type AccordionPanelProps = _AccordionPanelProps;
