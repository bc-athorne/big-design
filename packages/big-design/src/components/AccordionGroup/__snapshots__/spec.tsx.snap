// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders accordion group header 1`] = `
<h2
  class="styled__StyledH2-sc-tqnj75-2 styled__StyledH2-sc-1beu9nj-1 kEGvNd eqAvQy"
>
  Test Accordion
</h2>
`;
